### PR TITLE
Task03 Makar Selivanov SPbU

### PR DESCRIPTION
**Pull request by makselivanov**
<details><summary>Локальный вывод sum</summary>

<pre>
makar@mds-pc:build$ ./sum 0
CPU:     0.369145+-0.00892479 s
CPU:     270.896 millions/s
CPU OMP: 0.189399+-0.000809948 s
CPU OMP: 527.986 millions/s
OpenCL devices:
  Device #0: GPU. GeForce 940MX. Total memory: 983 Mb
  Device #1: CPU. Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz. Intel(R) Corporation. Total memory: 7852 Mb
  Device #2: GPU. Intel(R) HD Graphics 520 [0x1916]. Total memory: 6281 Mb
Using device #0: GPU. GeForce 940MX. Total memory: 983 Mb
GPU baseline:     0.0194048+-0.00087669 s
GPU baseline:     5153.36 millions/s
GPU bad loop:     0.033045+-0.000176115 s
GPU bad loop:     3026.18 millions/s
GPU coalesced loop:     0.0144413+-0.00145055 s
GPU coalesced loop:     6924.57 millions/s
GPU local memory:     0.144173+-0.000830386 s
GPU local memory:     693.612 millions/s
GPU tree:     0.048144+-0.000649678 s
GPU tree:     2077.1 millions/s
-----------------------------------------------------------------
makar@mds-pc:build$ ./sum 2
CPU:     0.38066+-0.0169368 s
CPU:     262.702 millions/s
CPU OMP: 0.185921+-0.00271915 s
CPU OMP: 537.863 millions/s
OpenCL devices:
  Device #0: GPU. GeForce 940MX. Total memory: 983 Mb
  Device #1: CPU. Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz. Intel(R) Corporation. Total memory: 7852 Mb
  Device #2: GPU. Intel(R) HD Graphics 520 [0x1916]. Total memory: 6281 Mb
Using device #2: GPU. Intel(R) HD Graphics 520 [0x1916]. Total memory: 6281 Mb
GPU baseline:     0.0285827+-0.00135276 s
GPU baseline:     3498.62 millions/s
GPU bad loop:     0.0492497+-0.000301903 s
GPU bad loop:     2030.47 millions/s
GPU coalesced loop:     0.0283292+-0.000214032 s
GPU coalesced loop:     3529.93 millions/s
GPU local memory:     0.170901+-0.0283067 s
GPU local memory:     585.136 millions/s
GPU tree:     0.0557437+-0.000630219 s
GPU tree:     1793.93 millions/s

</pre>

</details>

<details><summary>Вывод CI sum </summary>
<pre>
Run ./sum
CPU:     0.0771372+-8.58786e-05 s
CPU:     1296.39 millions/s
CPU OMP: 0.0421545+-9.48327e-05 s
CPU OMP: 2372.23 millions/s
OpenCL devices:
  Device #0: CPU. Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz. Intel(R) Corporation. Total memory: 6950 Mb
Using device #0: CPU. Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz. Intel(R) Corporation. Total memory: 6950 Mb
GPU baseline:     2.02147+-0.00699433 s
GPU baseline:     49.469 millions/s
GPU bad loop:     0.045206+-0.000186421 s
GPU bad loop:     2212.1 millions/s
GPU coalesced loop:     0.0396623+-8.06818e-05 s
GPU coalesced loop:     2521.28 millions/s
GPU local memory:     0.0708143+-0.000127764 s
GPU local memory:     1412.14 millions/s
GPU tree:     0.257743+-0.00088738 s
GPU tree:     387.983 millions/s
</pre>
</details>

<details><summary>Локальный вывод mandelbrot </summary>
<pre>
makar@mds-pc:build$ ./mandelbrot 0
OpenCL devices:
  Device #0: GPU. GeForce 940MX. Total memory: 983 Mb
  Device #1: CPU. Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz. Intel(R) Corporation. Total memory: 7852 Mb
  Device #2: GPU. Intel(R) HD Graphics 520 [0x1916]. Total memory: 6281 Mb
Using device #0: GPU. GeForce 940MX. Total memory: 983 Mb
CPU: 1.60593+-0.0112848 s
CPU: 6.2269 GFlops
    Real iterations fraction: 56.2638%
GPU: 1.17463+-0.000593216 s
GPU: 8.51334 GFlops
    Real iterations fraction: 56.2657%
GPU vs CPU average results difference: 0.942446%
-----------------------------------------------------------------
makar@mds-pc:build$ ./mandelbrot 2
OpenCL devices:
  Device #0: GPU. GeForce 940MX. Total memory: 983 Mb
  Device #1: CPU. Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz. Intel(R) Corporation. Total memory: 7852 Mb
  Device #2: GPU. Intel(R) HD Graphics 520 [0x1916]. Total memory: 6281 Mb
Using device #2: GPU. Intel(R) HD Graphics 520 [0x1916]. Total memory: 6281 Mb
CPU: 1.63232+-0.0454823 s
CPU: 6.12626 GFlops
    Real iterations fraction: 56.2638%
GPU: 0.870088+-0.00227273 s
GPU: 11.4931 GFlops
    Real iterations fraction: 56.2657%
GPU vs CPU average results difference: 0.942446%
</pre>
</details>

<details><summary>Вывод CI mandelbrot </summary>
<pre>
Run ./mandelbrot
OpenCL devices:
  Device #0: CPU. Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz. Intel(R) Corporation. Total memory: 6950 Mb
Using device #0: CPU. Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz. Intel(R) Corporation. Total memory: 6950 Mb
CPU: 1.87656+-0.00194899 s
CPU: 5.32889 GFlops
    Real iterations fraction: 56.2638%
GPU: 1.42174+-0.00141795 s
GPU: 7.03364 GFlops
    Real iterations fraction: 56.2657%
GPU vs CPU average results difference: 0.942446%
</pre>
</details>

### CODE REVIEW

**Comment by makselivanov** [comment]
Упс, забыл добавить main_sum

**Comment by makselivanov** [comment]
<details><summary>Локальный вывод sum</summary>

<pre>
makar@mds-pc:build$ ./sum 0
CPU:     0.369145+-0.00892479 s
CPU:     270.896 millions/s
CPU OMP: 0.189399+-0.000809948 s
CPU OMP: 527.986 millions/s
OpenCL devices:
  Device #0: GPU. GeForce 940MX. Total memory: 983 Mb
  Device #1: CPU. Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz. Intel(R) Corporation. Total memory: 7852 Mb
  Device #2: GPU. Intel(R) HD Graphics 520 [0x1916]. Total memory: 6281 Mb
Using device #0: GPU. GeForce 940MX. Total memory: 983 Mb
GPU baseline:     0.0194048+-0.00087669 s
GPU baseline:     5153.36 millions/s
GPU bad loop:     0.033045+-0.000176115 s
GPU bad loop:     3026.18 millions/s
GPU coalesced loop:     0.0144413+-0.00145055 s
GPU coalesced loop:     6924.57 millions/s
GPU local memory:     0.144173+-0.000830386 s
GPU local memory:     693.612 millions/s
GPU tree:     0.048144+-0.000649678 s
GPU tree:     2077.1 millions/s
-----------------------------------------------------------------
makar@mds-pc:build$ ./sum 2
CPU:     0.38066+-0.0169368 s
CPU:     262.702 millions/s
CPU OMP: 0.185921+-0.00271915 s
CPU OMP: 537.863 millions/s
OpenCL devices:
  Device #0: GPU. GeForce 940MX. Total memory: 983 Mb
  Device #1: CPU. Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz. Intel(R) Corporation. Total memory: 7852 Mb
  Device #2: GPU. Intel(R) HD Graphics 520 [0x1916]. Total memory: 6281 Mb
Using device #2: GPU. Intel(R) HD Graphics 520 [0x1916]. Total memory: 6281 Mb
GPU baseline:     0.0285827+-0.00135276 s
GPU baseline:     3498.62 millions/s
GPU bad loop:     0.0492497+-0.000301903 s
GPU bad loop:     2030.47 millions/s
GPU coalesced loop:     0.0283292+-0.000214032 s
GPU coalesced loop:     3529.93 millions/s
GPU local memory:     0.170901+-0.0283067 s
GPU local memory:     585.136 millions/s
GPU tree:     0.0557437+-0.000630219 s
GPU tree:     1793.93 millions/s

</pre>

</details>

<details><summary>Вывод CI sum </summary>
<pre>
Run ./sum
CPU:     0.0771372+-8.58786e-05 s
CPU:     1296.39 millions/s
CPU OMP: 0.0421545+-9.48327e-05 s
CPU OMP: 2372.23 millions/s
OpenCL devices:
  Device #0: CPU. Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz. Intel(R) Corporation. Total memory: 6950 Mb
Using device #0: CPU. Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz. Intel(R) Corporation. Total memory: 6950 Mb
GPU baseline:     2.02147+-0.00699433 s
GPU baseline:     49.469 millions/s
GPU bad loop:     0.045206+-0.000186421 s
GPU bad loop:     2212.1 millions/s
GPU coalesced loop:     0.0396623+-8.06818e-05 s
GPU coalesced loop:     2521.28 millions/s
GPU local memory:     0.0708143+-0.000127764 s
GPU local memory:     1412.14 millions/s
GPU tree:     0.257743+-0.00088738 s
GPU tree:     387.983 millions/s
</pre>
</details>

<details><summary>Локальный вывод mandelbrot </summary>
<pre>
makar@mds-pc:build$ ./mandelbrot 0
OpenCL devices:
  Device #0: GPU. GeForce 940MX. Total memory: 983 Mb
  Device #1: CPU. Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz. Intel(R) Corporation. Total memory: 7852 Mb
  Device #2: GPU. Intel(R) HD Graphics 520 [0x1916]. Total memory: 6281 Mb
Using device #0: GPU. GeForce 940MX. Total memory: 983 Mb
CPU: 1.60593+-0.0112848 s
CPU: 6.2269 GFlops
    Real iterations fraction: 56.2638%
GPU: 1.17463+-0.000593216 s
GPU: 8.51334 GFlops
    Real iterations fraction: 56.2657%
GPU vs CPU average results difference: 0.942446%
-----------------------------------------------------------------
makar@mds-pc:build$ ./mandelbrot 2
OpenCL devices:
  Device #0: GPU. GeForce 940MX. Total memory: 983 Mb
  Device #1: CPU. Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz. Intel(R) Corporation. Total memory: 7852 Mb
  Device #2: GPU. Intel(R) HD Graphics 520 [0x1916]. Total memory: 6281 Mb
Using device #2: GPU. Intel(R) HD Graphics 520 [0x1916]. Total memory: 6281 Mb
CPU: 1.63232+-0.0454823 s
CPU: 6.12626 GFlops
    Real iterations fraction: 56.2638%
GPU: 0.870088+-0.00227273 s
GPU: 11.4931 GFlops
    Real iterations fraction: 56.2657%
GPU vs CPU average results difference: 0.942446%
</pre>
</details>

<details><summary>Вывод CI mandelbrot </summary>
<pre>
Run ./mandelbrot
OpenCL devices:
  Device #0: CPU. Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz. Intel(R) Corporation. Total memory: 6950 Mb
Using device #0: CPU. Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz. Intel(R) Corporation. Total memory: 6950 Mb
CPU: 1.87656+-0.00194899 s
CPU: 5.32889 GFlops
    Real iterations fraction: 56.2638%
GPU: 1.42174+-0.00141795 s
GPU: 7.03364 GFlops
    Real iterations fraction: 56.2657%
GPU vs CPU average results difference: 0.942446%
</pre>
</details>

File: src/cl/sum.cl [hunk 1]
```diff
@@ -1 +1,111 @@
...
+
+__kernel void sum_loop_coalesced(
+        __global const unsigned int* a, unsigned int n,
+        __global       unsigned int* r)
+{
+    const unsigned int local_id = get_local_id(0);
+    const unsigned int group_id = get_group_id(0);
+    const unsigned int group_size = get_local_size(0);
+    unsigned int res = 0;
+    for (unsigned int i = 0; i < VALUES_PER_WORK_ITEM; ++i) {
```
Comment (vladimir-nazarenko) [inline src/cl/sum.cl:47]: Есть ещё прикольная идиома:
`for (int i = local_id_0; i < SZ; i += group_size)`
тогда 
`index = group_id * group_size * VALUES_PER_WORK_ITEM + i`

**Review by vladimir-nazarenko — COMMENTED** [review]
Супер! Задача зачтена, 10/10

**Pull request closed without merging**